### PR TITLE
Fix Metal Mipmap Behvior

### PR DESCRIPTION
**Pull request by cwfitzgerald**
**Checklist**

- [x] Run `cargo clippy`.
- [x] Run `RUSTFLAGS=--cfg=web_sys_unstable_apis cargo clippy --target wasm32-unknown-unknown` if applicable.
- [x] Add change to CHANGELOG.md. See simple instructions inside file.

**Description**

Couple issues I found when debugging a problem: 

- The mipmap filter was set to `NotMipmapped` (which unconditionally samples mip 0) when the LodClamp was None. This is incorrect, LodClamp of None just means there's no restrictions on the Lod.
- Mipmap Lod averaging was turned on. This has ramifications when neighboring pixels are `texelFetch`-ing texels of different mipmaps. This would potentially blend the mipmap levels in an unintended way.

Also resolves https://github.com/gfx-rs/wgpu/issues/3275.

**Testing**

Tested using rend3.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -113,6 +114,9 @@ By @teoxoy in [#3534](https://github.com/gfx-rs/wgpu/pull/3534)
 
 ### Bug Fixes
 
+#### Metal
+- Fix incorrect mipmap being sampled when using `MinLod <= 0.0` and `MaxLod >= 32.0` or when the fragment shaders in the same squad sample different Lods. By @cwfitzgerald in [#3610](https://github.com/gfx-rs/wgpu/pull/3610).
```
Comment (teoxoy) [inline CHANGELOG.md:118]:
```suggestion
- Fix incorrect mipmap being sampled when using `MinLod <= 0.0` and `MaxLod >= 32.0` or when the fragment shader samples different Lods in the same quad. By @cwfitzgerald in [#3610](https://github.com/gfx-rs/wgpu/pull/3610).
```

File: wgpu-hal/src/metal/device.rs [hunk 2]
```diff
@@ -409,16 +409,12 @@ impl crate::Device<super::Api> for super::Device {
...
-        let caps = &self.shared.private_caps;
         objc::rc::autoreleasepool(|| {
             let descriptor = metal::SamplerDescriptor::new();
 
             descriptor.set_min_filter(conv::map_filter_mode(desc.min_filter));
             descriptor.set_mag_filter(conv::map_filter_mode(desc.mag_filter));
             descriptor.set_mip_filter(match desc.mipmap_filter {
-                wgt::FilterMode::Nearest if desc.lod_clamp.is_none() => {
-                    metal::MTLSamplerMipFilter::NotMipmapped
-                }
```
Comment (teoxoy) [inline wgpu-hal/src/metal/device.rs:421]: I think we can still check if min and max are both 0 and take this branch.

**Review by teoxoy — CHANGES_REQUESTED** [review]
There are a few other things that don't look right with `lod_clamp` but could be addressed by a different PR.

- `lod_max_clamp` needs to default to `32` (right now it defaults to `f32::MAX`)
- `lod_clamp` is set to `None` if `lod_max_clamp > 32` which doesn't seem right
- `dx12` and `vulkan` then map a `lod_clamp: None` to `0.0..16.0` which also doesn't seem right; I think we should make `lod_clamp` required at the hal level

**Comment by cwfitzgerald** [comment]
K, went a little hog wild fixing up sampler stuff. I also brought aniso behavior inline with the spec and made the errors better.

File: wgpu/src/lib.rs [hunk 3]
```diff
@@ -1008,8 +1008,8 @@ pub struct SamplerDescriptor<'a> {
     pub lod_max_clamp: f32,
     /// If this is enabled, this is a comparison sampler using the given comparison function.
     pub compare: Option<CompareFunction>,
-    /// Valid values: 1, 2, 4, 8, and 16.
-    pub anisotropy_clamp: Option<NonZeroU8>,
+    /// Valid values between 1 and 16 inclusive. If this is not 1.0, all filter modes must be linear.
+    pub anisotropy_clamp: f32,
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply wgpu/src/lib.rs:1012]: Ah, misinterpreted `most implementations support` as a validation rule, will fix.

**Comment by teoxoy** [comment]
Ah, actually one remaining thing regarding the LOD changes, could you also update the Default impl of the sampler descriptors? `lod_max_clamp` needs to default to `32` but it currently defaults to `f32::MAX`.

I'm not sure why we have 2 types of `SamplerDescriptor` (one in wgpu-core and another in wgpu), they seem to have the same-ish data.

File: wgpu-hal/src/vulkan/device.rs [hunk 4]
```diff
@@ -1120,14 +1120,12 @@ impl crate::Device<super::Api> for super::Device {
                 .compare_op(conv::map_comparison(fun));
         }
 
-        if self
-            .shared
-            .downlevel_flags
-            .contains(wgt::DownlevelFlags::ANISOTROPIC_FILTERING)
-        {
+        if desc.anisotropy_clamp != 1 {
+            // We only enable the downlevel flag if supports 16x anisotropy,
```
Comment (teoxoy) [inline wgpu-hal/src/vulkan/device.rs:1124]:
```suggestion
            // We only enable anisotropy if it is supported,
```

Not sure if the wording is better (just a suggestion).

**Review by teoxoy — CHANGES_REQUESTED** [review]

**Review by teoxoy — APPROVED** [review]
Looks great!

> Someday I'll stop pushing commits to this damn PR

I think you can now :)

**Pull request merged**